WIP: Generalise Grid Synchronisation to Multi-Grid/Multi-Device

qin-yu opened the pull request:
Extracted `_get_handle()` and `_sync_handle()` for both grid group and multi-grid group. Also, added `cudaCGScope` to the same file as it is only used here. However, this cannot be used without a `cuLaunchCooperativeKernelMultiDevice()` CUDA API that will allow a single host thread to launch a kernel across multiple devices. This should be added to CUDAdrv.jl.

- [ ]  Add multi-device cooperative kernel launch to CUDAdrv
- [x]  Add `this_multi_grid()` and `sync_multi_grid()` to CUDAnative
- [ ]  Add tests for multi-device to CUDAnative

(Grid synchronisation was done in [Cooperative Kernel Launch #134](https://github.com/JuliaGPU/CUDAdrv.jl/pull/134) and [Grid Synchronization with Cooperative Groups #372](https://github.com/JuliaGPU/CUDAnative.jl/pull/372))

maleadt commented:
Your `llvmcall` invocations should eval the argument tuple, eg. `@eval Base.llvmcall($(decl, ir), ...)`, not the inner string (which you don't seem to be interpolating into anyway). See https://github.com/JuliaGPU/CUDAnative.jl/pull/379

maleadt commented on src/device/cuda/cooperative_groups.jl:
```diff
@@ -1,42 +1,71 @@
+# Both grid groups and multi-grid groups use the same intrinsic to get handle:
 if VERSION >= v"1.2.0-DEV.512"
-    @inline this_grid() = ccall("extern cudaCGGetIntrinsicHandle", llvmcall, Culonglong, (Cuint,), UInt32(1))
+    @inline _get_handle(cooperative_scope::cudaCGScope) =
```
maybe rename this to `cg_intrinsic_handle`; without namespacing this could clash or get confused with other handle functions

maleadt commented on src/device/cuda/cooperative_groups.jl:
```diff
@@ -1,42 +1,71 @@
-"""
-    this_grid()
+@enum cudaCGScope::Cuint begin
+      cudaCGScopeInvalid   = 0  # Invalid cooperative group scope
```
drop the redundant `cuda` on the values?

qin-yu replied:
Sure. (didn't learn from the CUDAapi PR 😂)

qin-yu committed:
Generalise Grid to Multi-Grid

Extracted `cg_intrinsic_handle()` and `cg_sync_handle()` for both grid 
group and 
multi-grid group. Also, added `cudaCGScope` to the same file as it is 
only used here.

qin-yu commented:
> Your `llvmcall` invocations should eval the argument tuple, eg. `@eval Base.llvmcall($(decl, ir), ...)`, not the inner string (which you don't seem to be interpolating into anyway).
> See #379 (that PR is probably also the source of conflict with `master` here).

Oh, yes, I made this PR before I read how you saved my `grid_sync()` in your bug fix PR. Changing it now.

qin-yu commented:
I didn't need it anymore so I stopped updating. Now I found that this was implemented in CUDA.jl(https://github.com/JuliaGPU/CUDA.jl)!
